pkg/controller/volume/persistentvolume/pv_controller.go

```diff
@@ -1341,13 +1424,44 @@ func (ctrl *PersistentVolumeController) provisionClaim(claim *v1.PersistentVolum
 	}
 	klog.V(4).Infof("provisionClaim[%s]: started", claimToClaimKey(claim))
 	opName := fmt.Sprintf("provision-%s[%s]", claimToClaimKey(claim), string(claim.UID))
-	startTime := time.Now()
-	ctrl.scheduleOperation(opName, func() error {
-		pluginName, err := ctrl.provisionClaimOperation(claim)
-		timeTaken := time.Since(startTime).Seconds()
-		metrics.RecordVolumeOperationMetric(pluginName, "provision", timeTaken, err)
-		return err
-	})
+	plugin, storageClass, err := ctrl.findProvisionablePlugin(claim)
+	if err != nil {
+		ctrl.eventRecorder.Event(claim, v1.EventTypeWarning, events.ProvisioningFailed, err.Error())
+		klog.Errorf("error finding provisioning plugin for claim %s: %v", claimToClaimKey(claim), err)
+		// failed to find the requested provisioning plugin, directly return err for now.
+		// controller will retry the provisioning in every syncUnboundClaim() call
+		// retain the original behavior of returning nil from provisionClaim call
+		return nil
+	}
+	// for CIS provisioning, latency metric will be reported after final binding happened
+	if plugin == nil || plugin.IsMigratedToCSI() {
+		// TEST case
+		ctrl.scheduleOperation(opName, func() error {
+			// create an opTimestamps without assigning provisionerName to get the current timestamp
+			opTimestamps := newOperationTimestamps("")
+			provisionerName, err := ctrl.provisionClaimOperationCSI(claim, plugin, storageClass)
+			opTimestamps.provisionerName = provisionerName
+			// adding an entry into operationTimestamps for metrics reporting if there exists no
+			// such record for current claim regardless whether the operation returned error or not
+			// add an entry to mark the provisioning request start time
+			claimKey := claimToClaimKey(claim)
+			if _, exists := ctrl.operationTimestamps.Get(claimKey); !exists {
+				ctrl.operationTimestamps.Add(claimKey, opTimestamps)
+			}
+			return err
+		})
+	} else /* plugin != nil && !plugin.IsMigratedToCSI() */ {
+		// record provision start time for operation metric reporting
+		startTime := time.Now()
+		ctrl.scheduleOperation(opName, func() error {
+			pluginName, err := ctrl.provisionClaimOperation(claim, plugin, storageClass)
+			// TODO: timeTaken here ONLY counts the first-try time and may not necessarily
+			//       represent the real end to end latency of provisioning
+			timeTaken := time.Since(startTime).Seconds()
+			metrics.RecordVolumeOperationMetric(pluginName, "provision", timeTaken, err)
```

Comment:
Are these metrics using same metric name as before? I think end-to-end metrics should use different name than current ones to avoid confusion.

Reply:
They are actually using the same name as before. The reason why I chose not to create new metric name is because this PR is a "bug-fix" like enhancement. The current implementation of latency metric for provisioning via CSI workflow was evaluated not in a proper way. Rather than creating a new metric, fill into the existing one would make existing monitoring dashboards/alerts seamlessly working without any change from end users and will stitch to in-tree metrics. However I do agree on your point that the naming could not be accurate. Happy to make a new name if you feel its more appropriate and there's no concern about backward compatibility though.

Reply:
I think we should actually create a new metric with a new name because the metric is going to have a completely different semantic now, and changing the semantic could break consumers that have already implemented alerts or SLOs around this metric.

Reply:
Another angle of looking at existing alerts/SLOs around this metric will be the correctness of them. As they (if any) have been built around an inaccurate (effectively close to zero for CSI workflow) implementation, end users who rely on the existing alerts might end up missing importation alerts or having their designed SLO missed without even notice. Is it legit to introduce a breaking change in this case instead of silently hiding by changing to use a different name?